Nishikawa Rp limiters (AIAA 2022-1473)

aeroamit opened the pull request:
## Proposed Changes
*Implementation of NISHIKAWA Rp Limiters (R3, R4, R5)
Reference - "New Unstructured-Grid Limiter Functions", Hiroaki Nishikawa, AIAA 2022-1473*
 


## Related Work
*1- Less dissipative in comparison to Venkatakrishnan limiter
  2- Can preserve higher order accuracy (3rd, 4th and 5th with R3, R4 and R5 limiters, probably future proof)
  3- May help in better convergence for certain cases with different K.*



## PR Checklist
*Put an X by all that apply. You can fill this out after submitting the PR. If you have any questions, don't hesitate to ask! We want to help. These are a guide for you to know what the reviewers will be looking for in your contribution.*

- [x] I am submitting my contribution to the develop branch.
- [x] My contribution generates no new compiler warnings (try with --warnlevel=3 when using meson).
- [x] My contribution is commented and consistent with SU2 style (https://su2code.github.io/docs_v7/Style-Guide/).
- [x] I have added a test case that demonstrates my contribution, if necessary.
- [x] I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp), if necessary.

aeroamit commented:
[Rp_limiter_ramp.pdf](https://github.com/su2code/SU2/files/11172822/Rp_limiter_ramp.pdf)
Above is a M=2.0 flow past a ramp in a channel case, comparing Venkat and Rp limiters. Have tried viscous (2D and 3D cases) as well.

bigfooted commented:
The results look very convincing, also nice to see such an improvement in convergence. Could you add the ramp case as a regression test to this PR (you could add them to parallel_regression.py)

bigfooted commented:
And can you also update the section on slope limiters in the documentation:
https://su2code.github.io/docs_v7/Slope-Limiters-and-Shock-Resolution/

aeroamit commented:
Sure @bigfooted , This is a  recent work by Dr. Nishikawa. Actually there are very few unstructured grid limiters in practice, Barth-Jespersen, Venkatakrishnan and its modifications, Michalek-Gooch (most of them are already implemented in SU2). I just saw the details and thought of implementing it. Also to note, it is tailored for vertex centered schemes not cell centered scheme, hence is apt for SU2. In original paper he showed tests with few inviscid cases. Recently He has revealed that it has been implemented in NASA CFD codes. 
Few things - 1- It is certainly less dissipative in comparison to VK 2- Convergence are similar. Rp limiters may take moderate number of more iterations (being less dissipative). 3- Convergence attributes also depends on many other settings - starting CFL, CFL ramping, convective schemes used etc. Sometimes all will stall.
Overall it's encouraging to try and make use of it in more and more cases. I tried it on multiple cases and got converged results (and yes Rp is comparitively less dissipative to Venkat limiter).
I will update the theory part and add the ramp case....

tbellosta commented:
Hi, I was about to open a pull request with the same exact implementation. I can help with the review and validation cases.
I am running test for an inviscid supersonic flow past a diamond airfoil and I am about to run some test on the RANS ONERA M6 wing.

tbellosta commented on SU2_CFD/include/limiters/CLimiterDetails.hpp:
```diff
@@ -159,6 +190,122 @@ struct CLimiterDetails<LIMITER::VENKATAKRISHNAN>
+  su2double epsp;
+
+  /*!
+   * \brief Store the reference lenght based eps^2 parameter,
```
eps^3

tbellosta commented on SU2_CFD/include/limiters/CLimiterDetails.hpp:
```diff
@@ -159,6 +190,122 @@ struct CLimiterDetails<LIMITER::VENKATAKRISHNAN>
+  su2double epsp;
+
+  /*!
+   * \brief Store the reference lenght based eps^2 parameter,
```
eps^4

tbellosta commented on SU2_CFD/include/limiters/CLimiterDetails.hpp:
```diff
@@ -159,6 +190,122 @@ struct CLimiterDetails<LIMITER::VENKATAKRISHNAN>
+  su2double epsp;
+
+  /*!
+   * \brief Store the reference lenght based eps^2 parameter,
```
eps^5

tbellosta reviewed (changes requested):
Implementation LGTM. 
Personally I would make const all that can be made const in the LimiterHelpers inlines, but I see you kept the style of the functions that were already there, so good for me.

tbellosta commented:
The first case I was running with the new limiters (Mach 2 invisccid flow around symmetrc diamond), showed similar convergence properties as the Venkatakrishnan limiter and slightly less dissipation as @aeroamit mentioned. The mesh I run was already quite fine, so no huge differences can be noticed but they are there.
![mach](https://user-images.githubusercontent.com/38865415/230571078-e5da281a-8d89-4f77-9e7b-746badbbc93c.png)

These are the  convergence history and a  slice of the solution (at k=0.05):
![conv0p05](https://user-images.githubusercontent.com/38865415/230574011-b30aee7c-d35e-4777-8a36-57111343a3fb.png)
![time0p05](https://user-images.githubusercontent.com/38865415/230574031-e7df6f3d-635b-4baf-9522-96fede31949e.png)
![full0p05](https://user-images.githubusercontent.com/38865415/230574171-d548babf-58d6-4f40-80d2-9a8326f4dc4f.png)

aeroamit commented:
Hi @tbellosta, Nice to see you. I have run many cases including ONEAR M6 RANS on a fine solve to wall mesh. Here are results -
[Onera_M6_runs.pdf](https://github.com/su2code/SU2/files/11177394/Onera_M6_runs.pdf)
Your run also confirms the observation I am reading from my cases.

tbellosta commented on SU2_CFD/include/limiters/CLimiterDetails.hpp:
```diff
@@ -87,6 +87,36 @@ struct LimiterHelpers
+    Type Dp = fabs(delta); Type Dm = fabs(proj);
+    if(Dp>(2.0*Dm)) {return 1.0;}
+    else {
+	Type y = Dp*Dp*Dp + epsp;
+	Type S3 = 4.0*Dm*Dm;	    
+ 	return ((y + Dp*S3) / (y + Dm*(delta*delta+S3)));}
```
indentation (there is also a check that complains about the use of tabs instead of spaces)

aeroamit replied:
done

tbellosta commented on SU2_CFD/include/limiters/CLimiterDetails.hpp:
```diff
@@ -87,6 +87,36 @@ struct LimiterHelpers
+    Type Dp = fabs(delta); Type Dm = fabs(proj);
+    if(Dp>(2.0*Dm)) {return 1.0;}
+    else {
+	Type y = Dp*Dp*Dp*Dp + epsp;
+        Type S4 = 2.0*Dm*(Dp*Dp-2.0*Dm*(Dp-2.0*Dm));	    
+        return ((y + Dp*S4) / (y + Dm*(delta*delta*delta+S4)));}
```
indentation

tbellosta commented on SU2_CFD/include/limiters/CLimiterDetails.hpp:
```diff
@@ -87,6 +87,36 @@ struct LimiterHelpers
+    Type Dp = fabs(delta); Type Dm = fabs(proj);
+    if(Dp>(2.0*Dm)) {return 1.0;}
+    else {
+	Type y = Dp*Dp*Dp*Dp*Dp + epsp;
+        Type S5 = 8.0*Dm*Dm*(Dp*Dp-2.0*Dm*(Dp-Dm));    
+        return ((y + Dp*S5) / (y + Dm*(delta*delta*delta*delta+S5)));}
```
indentation

tbellosta commented on SU2_CFD/include/limiters/CLimiterDetails.hpp:
```diff
@@ -140,6 +170,7 @@ struct CLimiterDetails<LIMITER::VENKATAKRISHNAN>
     su2double L = config.GetRefElemLength();
     su2double K = config.GetVenkat_LimiterCoeff();
     su2double eps1 = fabs(L*K);
+
```
I'd remove this to be consistent with the other limiters

pcarruscag commented on Common/include/option_structure.hpp:
```diff
@@ -895,6 +895,9 @@ static const MapType<std::string, ENUM_MATRIX_COLORING> MatrixColoring_Map = {
 enum class LIMITER {
   NONE                 , /*!< \brief No limiter. */
   VENKATAKRISHNAN      , /*!< \brief Slope limiter using Venkatakrisnan method (stencil formulation). */
+  NISHIKAWAR3          , /*!< \brief Slope limiter using NISHIKAWAR3 method (stencil formulation). */
```
For consistency with other names please separate with _
```suggestion
  NISHIKAWA_R3          , /*!< \brief Slope limiter using Nishikawa's R3 method (stencil formulation). */
```

pcarruscag commented on Common/src/CConfig.cpp:
```diff
@@ -6603,6 +6603,18 @@ void CConfig::SetOutput(SU2_COMPONENT val_software, unsigned short val_izone) {
           cout << "The reference element size is: " << RefElemLength << ". " << endl;
           break;
+        case LIMITER::NISHIKAWAR3:
+          cout << "NISHIKAWAR3 slope-limiting method, with constant: " << Venkat_LimiterCoeff << ".\n";
```
Same here (apply to R4 and R5 too please)
```suggestion
          cout << "Nishikawa's R3 slope-limiting method, with constant: " << Venkat_LimiterCoeff << ".\n";
```

pcarruscag commented on SU2_CFD/include/limiters/CLimiterDetails.hpp:
```diff
@@ -87,6 +87,36 @@ struct LimiterHelpers
+        Type y = Dp*Dp*Dp + epsp;
+        Type S3 = 4.0*Dm*Dm;
+        return ((y + Dp*S3) / (y + Dm*(delta*delta+S3)));}
+  }
```
For consistency with the other functions in this file:
```suggestion
  FORCEINLINE static Type r3Function(const Type& proj, const Type& delta, const Type& epsp)
  {
    Type Dp = fabs(delta);
    Type Dm = fabs(proj);
    if(Dp>(2.0*Dm)) return 1.0;
    Type y = pow(Dp, 3) + epsp;
    Type S3 = 4.0*Dm*Dm;
    return (y + Dp*S3) / (y + Dm*(delta*delta+S3));
  }
```

pcarruscag commented on SU2_CFD/include/limiters/CLimiterDetails.hpp:
```diff
@@ -87,6 +87,36 @@ struct LimiterHelpers
+    else {
+        Type y = Dp*Dp*Dp*Dp + epsp;
+        Type S4 = 2.0*Dm*(Dp*Dp-2.0*Dm*(Dp-2.0*Dm));
+        return ((y + Dp*S4) / (y + Dm*(delta*delta*delta+S4)));}
```
+ use pow for the higher powers please
```suggestion
        Type y = pow(Dp,4) + epsp;
        Type S4 = 2.0*Dm*(Dp*Dp-2.0*Dm*(Dp-2.0*Dm));
        return ((y + Dp*S4) / (y + Dm*(pow(delta,3)+S4)));}
```

pcarruscag commented on SU2_CFD/include/limiters/CLimiterDetails.hpp:
```diff
@@ -159,6 +189,122 @@ struct CLimiterDetails<LIMITER::VENKATAKRISHNAN>
+    su2double L = config.GetRefElemLength();
+    su2double K = config.GetVenkat_LimiterCoeff();
+    su2double eps1 = fabs(L*K);
+    epsp = max(eps1*eps1*eps1*eps1, LimiterHelpers<>::epsilon());
```
pow on these too

pcarruscag reviewed (commented):
This is great, please add one of those cases (maybe the M6) to the V&V page, we need contributions there https://su2code.github.io/vandv/home/

aeroamit commented:
Hi @pcarruscag , I have done the changes you suggested, please have a look.
Coming to https://su2code.github.io/vandv/home/  -  The ONERA M6 grid that I have used is somewhat big (still not what it can be), around 3 million mesh and 260 MB .su2 file. In V&V (I checked the cases), we have to put successively finer mesh and show grid independence, I don't have many such finer/coarse mesh (created suitably for grid independence). 
Probably I can take up any case from https://www.grc.nasa.gov/www/wind/valid/archive.html and try.  For example Sajben transonic diffuser or anything else. What's your opinion?

pcarruscag commented:
In that case maybe add the ramp case as a regression test, or maybe @tbellosta can contribute the diamond test as a regression.

As a separate effort, it would still be very nice to write up the M6 results for the V&V page, we can host large files in the SU2 foundations google drive. Having some reference results and settings known to work in that case would be very helpful for new folks.

pcarruscag reviewed (commented):
Implementation-wise all LGTM, thanks. Please add the regression test and then we can merge.
Also please don't forget the user guide page https://su2code.github.io/docs_v7/Slope-Limiters-and-Shock-Resolution/

aeroamit commented:
Sure, I will add the ramp case and update the documentation...

aeroamit commented:
I have added pull requests for ramp Test Case and documentation update (limiter portion) at respective places for review. When I should update the parallel_regression.py? Probably after addition of ramp case into develop? I have added one case few years back but not not able to recall the procedure...

pcarruscag commented:
The PR branch for the testcase is from your personal repo, so indeed we need to merge that first.

For branches in our repo you can change line 131 in the file .github/workflows/regression.yml to point to a different branch then develop

pcarruscag commented on TestCases/euler/ramp/inv_ramp.cfg:
```diff
@@ -0,0 +1,100 @@
+%
+SOLVER= EULER
+MATH_PROBLEM= DIRECT
+RESTART_SOL= NO
```
```suggestion
RESTART_SOL= YES
```

pcarruscag commented on TestCases/euler/ramp/inv_ramp.cfg:
```diff
@@ -0,0 +1,100 @@
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%                                                                              %
+% SU2 configuration file                                                       %
+% Case description: Supersonic flow over a ramp in a channel (regression)     %
+% Author: Thomas D. Economon, Amit Sachdeva                                                   %
```
```suggestion
% Case description: Supersonic flow over a ramp in a channel (regression)      %
% Author: Thomas D. Economon, Amit Sachdeva                                    %
```

pcarruscag commented on TestCases/euler/ramp/inv_ramp.cfg:
```diff
@@ -0,0 +1,100 @@
+CFL_NUMBER= 1.0
+CFL_ADAPT= YES
+CFL_ADAPT_PARAM= ( 0.9, 1.1, 1.0, 100.0, 0.8)
+RK_ALPHA_COEFF= ( 0.66667, 0.66667, 1.000000 )
```
```suggestion
```

pcarruscag commented on TestCases/euler/ramp/inv_ramp.cfg:
```diff
@@ -0,0 +1,100 @@
+NEWTON_KRYLOV= YES
+NEWTON_KRYLOV_IPARAM= (0, 1, 1) % n0, np, ft
+NEWTON_KRYLOV_DPARAM= (0.0, 1e-20, -2.0, 1e-5) % r0, tp, rf, e
+
+% -------------------------- MULTIGRID PARAMETERS -----------------------------%
+%
+MGLEVEL= 0
+MGCYCLE= V_CYCLE
+MG_PRE_SMOOTH= ( 1, 2, 3, 3 )
+MG_POST_SMOOTH= ( 0, 0, 0, 0 )
+MG_CORRECTION_SMOOTH= ( 0, 0, 0, 0 )
+MG_DAMP_RESTRICTION= 0.8
+MG_DAMP_PROLONGATION= 0.8
```
```suggestion
MGLEVEL= 0
```

pcarruscag commented on TestCases/parallel_regression.py:
```diff
@@ -257,6 +257,14 @@ def main():
+    ramp.cfg_file = "inv_ramp.cfg"
+    ramp.test_iter = 10
+    ramp.test_vals = [-13.399623, -7.788893, -0.081064, 0.056474] #last 4 columns
+    ramp.append(ramp)
```
```suggestion
    test_list.append(ramp)
```

aeroamit commented:
@pcarruscag , kindly change the  name of restart_flow.dat to solution_flow.dat for Inviscid supersonic flow past a ramp in a channel #122 pull request (merged) case. That's why regression failed this time (could not find restart file solution_flow.dat).

aeroamit commented:
> @pcarruscag , kindly change the name of restart_flow.dat to solution_flow.dat for Inviscid supersonic flow past a ramp in a channel #122 pull request (merged) case. That's why regression failed this time (could not find restart file solution_flow.dat).

Or perhaps I can change the restart file name to restart_flow.dat itself for passing the regression test (may not be a good practice)?

pcarruscag commented:
Its fine to change the name in the config file
